cdap-app-templates/cdap-program-report/src/main/java/co/cask/cdap/report/proto/ReportGenerationRequest.java

```diff
@@ -0,0 +1,407 @@
+
+/**
+ * Represents a request to generate a program run report in an HTTP request.
+ */
+public class ReportGenerationRequest {
+  private final Long start;
+  private final Long end;
+  private final List<String> fields;
+  @Nullable
+  private final List<Sort> sort;
+  @Nullable
+  private final List<Filter> filters;
+
+  public ReportGenerationRequest(Long start, Long end, List<String> fields, @Nullable List<Sort> sort,
+                                 @Nullable List<Filter> filters) {
+    this.start = start;
+    this.end = end;
+    this.fields = fields;
+    this.sort = sort;
+    this.filters = filters;
+    this.validate();
+  }
+
+  /**
+   * @return the start of the time range within which the report is generated. All program runs in the report must be
+   *         still running at {@code start}, or with end time not earlier than {@code start}.
+   */
+  public Long getStart() {
+    return start;
+  }
+
+  /**
+   * @return the end of the time range within which the report is generated. All program runs in the report must be
+   *         still start before {@code end}.
+   */
+  public Long getEnd() {
+    return end;
+  }
+
+  /**
+   * @return names of the fields to be included in the final report. Must be valid fields from {@link ReportField}.
+   */
+  public List<String> getFields() {
+    return fields;
+  }
+
+  /**
+   * @return the field to sort the report by. Currently only support a single field.
+   */
+  @Nullable
+  public List<Sort> getSort() {
+    return sort;
+  }
+
+  /**
+   * @return the filters that must be satisfied for every record in the report.
+   */
+  @Nullable
+  public List<Filter> getFilters() {
+    return filters;
+  }
+
+  /**
+   * Validates this {@link ReportGenerationRequest}
+   *
+   * @throws IllegalArgumentException if this request is not valid.
+   */
+  public void validate() {
+    List<String> errors = new ArrayList<>();
+    if (start == null) {
+      errors.add("'start' must be specified.");
+    }
+    if (end == null) {
+      errors.add("'end' must be specified.");
+    }
+    if (start >= end) {
+      errors.add("'start' must be smaller than 'end'.");
+    }
+    if (fields == null || fields.isEmpty()) {
+      errors.add("'fields' must be specified.");
+    } else {
+      errors.addAll(fields.stream().filter(field -> !ReportField.isValidField(field))
+                      .map(field -> String.format("Invalid field name '%s' in fields. Field name must be one of: [%s]",
+                                                  field, String.join(", ", ReportField.FIELD_NAME_MAP.keySet())))
+                      .collect(Collectors.toList()));
+    }
+    if (filters != null) {
+      filters.stream().map(Field::getErrors).forEach(errors::addAll);
+    }
+    if (sort != null) {
+      if (sort.size() > 1) {
+        errors.add("Currently only one field is supported in sort.");
+      }
+      sort.stream().map(Field::getErrors).forEach(errors::addAll);
```

Comment:
shouldn't you be doing `sort::getErrors` ? similar question for filter getErrors